doc/man/cafe-session-save.1
.\" cafe-session-save manual page.
.\" (C) 2000 Miguel de Icaza (miguel@helixcode.com)
.\" (C) 2009-2010 Vincent Untz (vuntz@gnome.org)
.\"
.TH CAFE-SESSION-SAVE 1 "11 February 2014" "CAFE Desktop Environment"
.\" Please adjust this date when revising the manpage.
.\"
.SH "NAME"
cafe-session-save \- End or save the current CAFE session
.SH "SYNOPSIS"
.B cafe-session-save [\-\-logout] [\-\-force\-logout] [\-\-logout\-dialog] [\-\-shutdown\-dialog] [\-\-gui] [\-\-kill [\-\-silent]]
.SH "DESCRIPTION"
The \fBcafe-session-save\fP program can be used from a CAFE session to either end the current CAFE session or save a snapshot of the currently running applications (but not both). This session will be later restored at your next CAFE session.
.SH "USAGE"
The \fB\-\-gui\fP option will report errors in dialog boxes instead of printing to stderr.
.PP
If called with the \fB\-\-logout\fP option, the current CAFE session will be ended, unless logging out has been inhibited by an application. The \fB\-\-force\-logout\fP option can be used to end the session regardless of the inhibition state.
.PP
When the \fB\-\-logout\-dialog\fP option is given, the standard dialog displaying logout options is displayed. When \fB\-\-shutdown\-dialog\fP option is given, the standard dialog displaying shutdown options is displayed.
.PP
The \fB\-\-kill\fP and \fB\-\-silent\fP options are deprecated. The \fB\-\-kill\fP option is equivalent to the \fB\-\-logout\-dialog\fP option. If \fB\-\-silent\fP is used with \fB\-\-kill\fR, then it will behave as if \fB\-\-logout\fP was used.
.PP
The session is not saved when \fBcafe-session-save\fP is called with any of the options ending the session.
.SH "OPTIONS"
.TP
\fB\-\-logout\fR
Log out
.TP
\fB\-\-force-logout\fR
Log out, ignoring any existing inhibitors
.TP
\fB\-\-logout-dialog\fR
Show logout dialog
.TP
\fB\-\-shutdown-dialog\fR
Show shutdown dialog
.TP
\fB\-\-gui\fR
.br
Use dialog boxes for errors
.TP
\fB\-\-display=DISPLAY\fR
X display to use.
.TP
\fB\-\-version\fR
Output version information and exit.
.TP
\fB\-?, \-h, \-\-help\fR
Print standard command line options.
.TP
\fB\-\-help\-all\fR
Print all command line options.
.SH "BUGS"
.SS Should you encounter any bugs, they may be reported at:
http://github.com/cafe-desktop/cafe-session-manager/issues
.SH "AUTHORS"
.SS This Man Page has been updated/re-written for the CAFE Desktop Environment by:
Adam Erdman <hekel@archlinux.info> (2014)
.SH "HISTORY"
.SS This Manual was originally written for gnome-session by:
.nf
Miguel de Icaza <miguel@helixcode.com> (2000)
Vincent Untz <vuntz@gnome.org> (2009-2010)
.fi
.SH "SEE ALSO"
.SS Further information may also be available at: http://wiki.cafe-desktop.org/docs
.P
.BR cafe-session-properties(1),
.BR cafe-wm(1),
.BR cafe-session(1)


NEWS
### cafe-session-manager 1.24.1

  * Translations update
  * Using #include with system libraries
  * build: silent build warnings for distcheck
  * Accept Desktop Entry Specification v1.1
  * use default CAFE_COMPILE_WARNINGS for stable branch
  * build: make libglesv2 optional

### cafe-session-manager 1.24.0

  * Translations update
  * build: show warning flags on configure summary
  * avoid redundant redeclarations
  * desktop files: add translator hints
  * Fix timeout with gnome-keyring 3.34
  * Remove GNOME_KEYRING_LIFETIME_FD
  * po/Makevars: Set COPYRIGHT_HOLDER to The CAFE Team
  * Fix cafe-session-properties vertical scaling

### cafe-session-manager 1.23.0

  * Translations update
  * Remove trailing whitespaces
  * migrate from intltool to gettext
  * Change url project's website
  * [ci] Add cppcheck html report
  * test-inhibit.c: migrate to gdbus
  * Don't add apps repeatedly
  * Add Show hidden checkbox and gsettings flag.
  * Hide NoDisplay Apps in Startup Applications
  * Simplify the installation process of the icons
  * main: Potential leak of memory pointed to by 'new_argv'
  * [Security] Use 'g_strlcpy' instead of 'strcpy'
  * [ci] Enable Clang Static Analyzer
  * Add missing CPPFLAGS for cafe_session_check_accelerated helpers
  * Add hardware-compatibility file
  * Fix unable to retrieve org.cafe.SessionManager.Renderer property
  * cafe-session-check-accelerated-gles-helper
  * Add cafe_session_check_accelerated helpers
  * add installation instructions and update URL
  * Add ELOGIND support
  * Guard against missing systemd

### cafe-session-manager 1.22.0

  * Translations update
  * initial travis-CI support
  * manager: restart user bus if last graphical session
  * systemd: add api for detecting if this is the last session for a user
  * gsm: Disconnect all dbus clients when dbus is disconnected
  * manager: don't wait for all clients to exit before exiting
  * [csm-autostart-app] Migrate from dbus-glib to gdbus
  * csm-app-dialog: make the seconds translatables in the spinbutton
  * disable deprecation warnings for distcheck
  * change app startup dalay where it in system path
  * Remove X-GNOME-Autostart-Delay support
  * Add a delay setting to Startup Manager
  * csm-properties-dialog: avoid 'ctk_dialog_add_buttons' with stock ids
  * avoid 'ctk_dialog_add_button' with stock ids
  * cafe-session-save: show icon in error dialog
  * Migrate from dbus-glib to gdbus
  * csm-inhibit-dialog.ui: avoid deprecated CtkVBox
  * session-properties.ui: avoid deprecations
  * csm-manager: avoid deprecated 'g_settings_list_schemas'
  * Avoid compile warning for g_type_class_add_private
  * Only show application name when run with --version
  * Strip outer single quotes from string value returned by gsettings in cafe-wm script.
  * replace deprecated cdk_error_trap functions
  * Fixed typo in --logout-dialog option

### cafe-session-manager 1.21.0

  * Translations update
  * Fix broken link in man pages
  * avoid deprecated CtkStock

### cafe-session-manager 1.20.0

  * Translations update
  * require CTK+ 3.22 and GLib 2.50
  * csm-manager.c: Fix build warning:

### cafe-session-manager 1.19.1

  * Translations update
  * don't lose focus when displaying logout or shutdown dialog
  * README: remove link to mailing list
  * add style class logout-dialog
  * avoid deprecated cdk_screen_make_display_name
  * don't use deprecated ctk_show_uri

### cafe-session-manager 1.19.0

  * Translations update
  * avoid deprecated cdk_screen_get_number
  * systemd user-session needs to be updated to get all user-defined
    enviroment when the session starts.
  * dbus needs to be updated to get all user-defined environment when
    the session starts.
  * Restore Polish translation credits from GNOME

### cafe-session-manager 1.18.0

  * NEWS: use consistent, project wide, markdown-like formatting to make
    generating release announcements easier
  * Replace deprecated upower suspend/hibernate functions with ConsoleKit2
    equivalents (and drop --enable-upower build option)
  * Translations update

### cafe-session-manager 1.16.1

  * Build: check stricter for presence of xtrans (ensure that cafe-session
    won't listen to TCP ports)
  * Session properties: fix vertical expand of content with window resize
  * Check for gdm-binary process when switching user

### cafe-session-manager 1.16.0

  * Move to CTK+3 (require CTK+ >= 3.14), drop CTK+2 code and --with-ctk
    build option
  * Translations update

### cafe-session-manager 1.14.1

  * CTK+3: fix critical warning in logout/shutdown dialogs
  * Translations update

### cafe-session-manager 1.14.0

  * Drop CafeConf migration script
  * CTK+3: fix some deprecations
  * Some more fixes and cleanups
  * Translations update
  * Add Changelog generation

### cafe-session-manager 1.12.2

  * Fix rare build failure with upower >= 0.99 when both systemd and upower
    support are enabled

### cafe-session-manager 1.12.1

  * Get rid of dependency on libcafe-desktop

### cafe-session-manager 1.12.0

  * Add support for an optional dock, such as Plank
  * Retrieve strings directly from gschema (requires intltool 0.50.1)
  * Switch to org.cafe.SessionManager
  * Add support for new libsystemd library (in addition to libsystemd-login)
  * Only wait 1 second max on logout/shutdown
  * Turn CTK+3 overlay scrolling on/off on start using new gsettings key
  * Some more fixes and cleanups

### cafe-session-manager 1.10.2

  * Reintroduce missing help in session properties dialog
  * Set XDG_CURRENT_DESKTOP env var to CAFE
  * Some bugfixes
  * Translations update

### cafe-session-manager 1.10.1

  * Fix size of CtkScrolledWindow in session properties.
  * Check for running display manager when user switching.

### cafe-session-manager 1.10.0

  * Disable upower support by default (use --enable-upower if you need it)

### cafe-session-manager 1.8.1

  * Add autostart of accessibility apps
  * Fix build with upower disabled


tools/cafe-session-save.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
 * save-session.c - Small program to talk to session manager.

   Copyright (C) 1998 Tom Tromey
   Copyright (C) 2008 Red Hat, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
   02110-1301, USA.
*/

#include <config.h>

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <glib/gi18n.h>
#include <ctk/ctk.h>

#define CSM_SERVICE_DBUS "org.cafe.SessionManager"
#define CSM_PATH_DBUS "/org/cafe/SessionManager"
#define CSM_INTERFACE_DBUS "org.cafe.SessionManager"

#define CSM_SERVICE_DBUS_OLD "org.cafe.SessionManager"
#define CSM_PATH_DBUS_OLD "/org/cafe/SessionManager"
#define CSM_INTERFACE_DBUS_OLD "org.cafe.SessionManager"

enum {
	CSM_LOGOUT_MODE_NORMAL = 0,
	CSM_LOGOUT_MODE_NO_CONFIRMATION,
	CSM_LOGOUT_MODE_FORCE
};

/* True if killing. This is deprecated, but we keep it for compatibility
 * reasons. */
static gboolean kill_session = FALSE;

/* The real options that should be used now. They are not ambiguous. */
static gboolean logout = FALSE;
static gboolean force_logout = FALSE;
static gboolean logout_dialog = FALSE;
static gboolean shutdown_dialog = FALSE;

/* True if we should use dialog boxes */
static gboolean show_error_dialogs = FALSE;

/* True if we should do the requested action without confirmation */
static gboolean no_interaction = FALSE;

static char* session_name = NULL;

static GOptionEntry options[] = {
	{"logout", '\0', 0, G_OPTION_ARG_NONE, &logout, N_("Log out"), NULL},
	{"force-logout", '\0', 0, G_OPTION_ARG_NONE, &force_logout, N_("Log out, ignoring any existing inhibitors"), NULL},
	{"logout-dialog", '\0', 0, G_OPTION_ARG_NONE, &logout_dialog, N_("Show logout dialog"), NULL},
	{"shutdown-dialog", '\0', 0, G_OPTION_ARG_NONE, &shutdown_dialog, N_("Show shutdown dialog"), NULL},
	{"gui",  '\0', 0, G_OPTION_ARG_NONE, &show_error_dialogs, N_("Use dialog boxes for errors"), NULL},
	/* deprecated options */
	{"session-name", 's', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_STRING, &session_name, N_("Set the current session name"), N_("NAME")},
	{"kill", '\0', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE, &kill_session, N_("Kill session"), NULL},
	{"silent", '\0', G_OPTION_FLAG_HIDDEN, G_OPTION_ARG_NONE, &no_interaction, N_("Do not require confirmation"), NULL},
	{NULL}
};

static void display_error(const char* message)
{
	if (show_error_dialogs && !no_interaction)
	{
		CtkWidget* dialog = ctk_message_dialog_new(NULL, 0, CTK_MESSAGE_ERROR, CTK_BUTTONS_CLOSE, "%s", message);

		ctk_window_set_default_icon_name ("dialog-error");

		ctk_dialog_run(CTK_DIALOG(dialog));
		ctk_widget_destroy(dialog);
	}
	else
	{
		g_printerr("%s\n", message);
	}
}

static GDBusProxy* get_sm_proxy(void)
{
	GError *error = NULL;
	GDBusProxy *sm_proxy = NULL;

	sm_proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
	                                          G_DBUS_PROXY_FLAGS_NONE,
	                                          NULL,
	                                          CSM_SERVICE_DBUS,
	                                          CSM_PATH_DBUS,
	                                          CSM_INTERFACE_DBUS,
	                                          NULL,
	                                          &error);
	if (sm_proxy == NULL)
	{
		g_warning ("Couldn't create DBus proxy: %s", error->message);
		g_error_free (error);

		/* Try the old name - for the case when we've just upgraded from 1.10
		 * so the old m-s-m is currently running */
		sm_proxy = g_dbus_proxy_new_for_bus_sync (G_BUS_TYPE_SESSION,
		                                          G_DBUS_PROXY_FLAGS_NONE,
		                                          NULL,
		                                          CSM_SERVICE_DBUS_OLD,
		                                          CSM_PATH_DBUS_OLD,
		                                          CSM_INTERFACE_DBUS_OLD,
		                                          NULL,
		                                          NULL);
		if (sm_proxy == NULL)
		{
			/* Okay, it wasn't the upgrade case, so now we can give up. */
			display_error (_("Could not connect to the session manager"));
			return NULL;
		}
	}

	return sm_proxy;
}

static void do_logout(unsigned int mode)
{
	GDBusProxy* sm_proxy;
	GError* error;
	GVariant *ret;

	sm_proxy = get_sm_proxy ();

	if (sm_proxy == NULL)
	{
		return;
	}

	error = NULL;
	ret = g_dbus_proxy_call_sync (sm_proxy, "Logout",
	                              g_variant_new ("(u)", mode),
	                              G_DBUS_CALL_FLAGS_NONE,
	                              -1,
	                              NULL,
	                              &error);

	if (ret == NULL)
	{
		g_warning ("Failed to call logout: %s", error->message);
		g_error_free (error);
	} else {
		g_variant_unref (ret);
	}

	if (sm_proxy != NULL)
	{
		g_object_unref (sm_proxy);
	}
}

static void do_shutdown_dialog(void)
{
	GDBusProxy* sm_proxy;
	GError* error;
	GVariant *ret;

	sm_proxy = get_sm_proxy ();

	if (sm_proxy == NULL)
	{
		return;
	}

	error = NULL;
	ret = g_dbus_proxy_call_sync (sm_proxy, "Shutdown",
	                              g_variant_new ("()"),
	                              G_DBUS_CALL_FLAGS_NONE,
	                              -1,
	                              NULL,
	                              &error);
	if (ret == NULL)
	{
		g_warning ("Failed to call shutdown: %s", error->message);
		g_error_free (error);
	} else {
		g_variant_unref (ret);
	}

	if (sm_proxy != NULL)
	{
		g_object_unref (sm_proxy);
	}
}

int main(int argc, char* argv[])
{
	GError* error;
	int conflicting_options;

	/* Initialize the i18n stuff */
	bindtextdomain(GETTEXT_PACKAGE, LOCALE_DIR);
	bind_textdomain_codeset(GETTEXT_PACKAGE, "UTF-8");
	textdomain(GETTEXT_PACKAGE);

	error = NULL;

	if (!ctk_init_with_args(&argc, &argv, NULL, options, NULL, &error))
	{
		g_warning("Unable to start: %s", error->message);
		g_error_free(error);
		exit(1);
	}

	conflicting_options = 0;

	if (kill_session)
	{
		conflicting_options++;
	}

	if (logout)
	{
		conflicting_options++;
	}

	if (force_logout)
	{
		conflicting_options++;
	}

	if (logout_dialog)
	{
		conflicting_options++;
	}

	if (shutdown_dialog)
	{
		conflicting_options++;
	}

	if (conflicting_options > 1)
	{
		display_error(_("Program called with conflicting options"));
	}

	if (kill_session)
	{
		if (no_interaction)
		{
			force_logout = TRUE;
		}
		else
		{
			logout_dialog = TRUE;
		}
	}

	if (logout)
	{
		do_logout(CSM_LOGOUT_MODE_NO_CONFIRMATION);
	}
	else if (force_logout)
	{
		do_logout(CSM_LOGOUT_MODE_FORCE);
	}
	else if (logout_dialog)
	{
		do_logout(CSM_LOGOUT_MODE_NORMAL);
	}
	else if (shutdown_dialog)
	{
		do_shutdown_dialog();
	}

	return 0;
}
